components/script/dom/promisedebugging.rs

```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::inheritance::Castable;
+use dom::bindings::root::DomRoot;
+use dom::event::{Event, EventBubbles, EventCancelable};
+use dom::eventtarget::EventTarget;
+use dom::globalscope::GlobalScope;
+use dom::promise::Promise;
+use dom::promiserejectionevent::PromiseRejectionEvent;
+use js::jsapi::{Heap, JSObject, HandleObject as RawHandleObject};
+use js::rust::wrappers::GetPromiseResult;
+use std::cell::Cell;
+
+pub struct PromiseDebugging {
+  pub id_predix: Cell<String>,
+}
+
+impl PromiseDebugging {
+    #[allow(unsafe_code)]
+    pub fn AddUncaughtRejection(global: DomRoot<GlobalScope>, promise: RawHandleObject) {
+        global.set_uncaught_rejections(promise);
+
+        unsafe {
+            PromiseDebugging::flush_uncaught_rejections_internal(global);
```

Comment:
@jdm cc @nox 

In Gecko, `flush uncaught rejections internal` will only be triggered in [`FlushSync`](https://searchfox.org/mozilla-central/source/dom/promise/PromiseDebugging.cpp#60-63).

So, I'd like to confirm if I'm correct about doing the similar implementation in Servo like following:
- Make the received uncaught rejection be queued into `script thread` with `task` which should be the step 4 in [notify-about-rejected-promises spec](https://html.spec.whatwg.org/multipage/webappapis.html#notify-about-rejected-promises) instead of running the `flush_uncaught_rejections_internal` directly when `AddUncaughtRejection` executed

Reply:
@jdm ping for asking question, thanks!

Reply:
I will try to use `task` macro to handle `Queue a task` for unhandled rejection.

Reply:
Sorry for not responding. Yes, queuing a task using the task macro is the right way to implement step 4. However, that task should only be queued once per run of the event loop, and I believe AddUncaughtRejection can run an arbitrary number of times per run of the event loop. You'll want to invoke the [notify about rejected promises](https://html.spec.whatwg.org/multipage/webappapis.html#notify-about-rejected-promises) algorithm as part of the [perform a microtask checkpoint](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model:notify-about-rejected-promises) algorithm.